source/mir/cv/imgproc/imfilter.d

```diff
@@ -0,0 +1,469 @@
+
+    scope (exit)
+    {
+        deallocate(tempbuf);
+    }
+
+    auto input = inputbuf.sliced(rows, cols);
+    auto temp = tempbuf.sliced(rows, cols);
+    auto output = outputbuf.sliced(rows, cols);
+    auto hm = hmask.sliced(ksize);
+    auto vm = vmask.sliced(ksize);
+
+    if (avx) // add version (compile time flag if avx should be supported)
+    {
+        inner_filtering_impl!(AVX!T, apply_horizontal_kernel_simd!(AVX!T), apply_vertical_kernel_simd!(AVX!T))
+            (input, temp, hm, vm, output);
+    }
+    else if (sse)
+    {
+        inner_filtering_impl!(SSE!T, apply_horizontal_kernel_simd!(SSE!T), apply_vertical_kernel_simd!(SSE!T))
+            (input, temp, hm, vm, output);
+    }
+    else
+    {
+        switch (ksize) {
+        case 3:
+            inner_filtering_impl!(Non_SIMD!T, apply_horizontal_kernel_3!T, apply_vertical_kernel_3!T)
+                (input, temp, hm, vm, output);
+            break;
+        case 5:
+            inner_filtering_impl!(Non_SIMD!T, apply_horizontal_kernel_5!T, apply_vertical_kernel_5!T)
+                (input, temp, hm, vm, output);
+            break;
+        case 7:
+            inner_filtering_impl!(Non_SIMD!T, apply_horizontal_kernel_7!T, apply_vertical_kernel_7!T)
+                (input, temp, hm, vm, output);
+            break;
+        default:
+            inner_filtering_impl!(Non_SIMD!T, apply_horizontal_kernel!T, apply_vertical_kernel!T)
+                (input, temp, hm, vm, output);
+        }
+    }
+
+    if (border_handling != BORDER_REPLICATE)
+    {
+        return ERROR_INVALID_BORDER_METHOD;
+    }
+
+    borders_replicate_impl(output, ksize);
+
+    return 0;
+}
+
+pragma(inline, true)
+void inner_filtering_impl
+(
+    alias InstructionSet,
+    alias hkernel,
+    alias vkernel
+)
+(
+    Slice!(Contiguous, [2], InstructionSet.Scalar*) input,
+    Slice!(Contiguous, [2], InstructionSet.Scalar*) temp,
+    Slice!(Contiguous, [1], InstructionSet.Scalar*) hmask,
+    Slice!(Contiguous, [1], InstructionSet.Scalar*) vmask,
+    Slice!(Contiguous, [2], InstructionSet.Scalar*) output
+)
+in
+{
+    assert(input.shape == temp.shape, "Incompatible input and temporary buffer slice");
+    assert(temp.shape == output.shape, "Incompatible input and output buffer slice");
+}
+body
+{
+    alias T = InstructionSet.Scalar;
+    alias V = InstructionSet.Vector;
+    immutable velems = InstructionSet.elementCount;
+    immutable tbytes = T.sizeof;
+    immutable vbytes = V.sizeof;
+    immutable ksize = hmask.length;
+    immutable rows = input.length!0;
+    immutable cols = input.length!1;
+
+    static if (Is_SIMD!InstructionSet)
+    {
+        // It this is SIMD instructions set, allocate vectors for kernels
+
+        auto hk = cast(V[])alignedAllocate(ksize * vbytes, 16); // horizontal simd kernel
+        auto vk = cast(V[])alignedAllocate(ksize * vbytes, 16); // vertical simd kernel
```

Comment:
You can multiple vector by scalar. No need to allocate memory and fill it with vectors.